Pylint causes a crash

Not sure what's causing an error here, I ran `repobee repos clone --sf ../students.tsv -o datintro20 --discover-repos --tb`:
```
Executing post_clone hooks:  51%|████████▏       | 19/37 [00:00<00:00, 184.95it/s]
Running pylint on /tmp/tmpep8vtouy/6b2f8909f58397c2f0ae7ae4bb7cd95772198f14/L1.py
Executing post_clone hooks:  73%|████████████▍    | 27/37 [00:00<00:00, 47.01it/s]
[ERROR] too many values to unpack (expected 2)
[ERROR] Critical exception      
Traceback (most recent call last):                                                
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/_repobee/main.py", li
ne 152, in main                          
    _repobee.cli.dispatch.dispatch_command(
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/_repobee/cli/dispatch
.py", line 55, in dispatch_command
    dispatch_table[category](args, config_file, api) or hook_results
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/_repobee/cli/dispatch
.py", line 90, in _dispatch_repos_command 
    return command.clone_repos(args.repos, api)
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/_repobee/command/repo
s.py", line 326, in clone_repos
    return plugin.execute_clone_tasks(local_repos_progress, api)
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/_repobee/plugin.py", 
line 387, in execute_clone_tasks
    return _execute_tasks(repos, plug.manager.hook.post_clone, api, cwd)
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/_repobee/plugin.py", 
line 426, in _execute_tasks
    for result in hook_function(repo=repo, api=api):
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/pluggy/hooks.py", lin
e 286, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/pluggy/manager.py", l
ine 93, in _hookexec 
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/pluggy/manager.py", l
ine 84, in <lambda>
    self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/pluggy/callers.py", l
ine 208, in _multicall
    return outcome.get_result()
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/pluggy/callers.py", l
ine 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/pluggy/callers.py", l
ine 187, in _multicall
    res = hook_impl.function(*args)
  File "/home/dbosk/.repobee/env/lib/python3.8/site-packages/_repobee/ext/pylint.p
y", line 47, in post_clone
    status, msg = _pylint(python_files)
ValueError: too many values to unpack (expected 2)
```

--- a/src/_repobee/ext/pylint.py
+++ b/src/_repobee/ext/pylint.py
@@ -17,7 +17,10 @@ files in a repo, and runs pylint on them, storing the results in files named
 
 import subprocess
 import pathlib
-from typing import Tuple, Union, Iterable
+import shlex
+import sys
+import dataclasses
+from typing import List
 
 
 import repobee_plug as plug
@@ -27,8 +30,15 @@ PLUGIN_DESCRIPTION = "Runs pylint on student repos after cloning"
 SECTION = "pylint"
 
 
+@dataclasses.dataclass(frozen=True)
+class _PylintResult:
+    path: pathlib.Path
+    output: str
+    errored: bool
+
+
 @plug.repobee_hook
-def post_clone(repo: plug.StudentRepo, api: plug.PlatformAPI):
+def post_clone(repo: plug.StudentRepo, api: plug.PlatformAPI) -> plug.Result:
     """Run pylint on all Python files in a repo.
 
     Args:
@@ -37,40 +47,62 @@ def post_clone(repo: plug.StudentRepo, api: plug.PlatformAPI):
     Returns:
         a plug.Result specifying the outcome.
     """
-    path = repo.path
-    python_files = list(path.rglob("*.py"))
-
-    if not python_files:
+    lint_results = _pylint(repo.path)
+    if not lint_results:
         msg = "no .py files found"
         return plug.Result(SECTION, plug.Status.WARNING, msg)
 
-    status, msg = _pylint(python_files)
-    return plug.Result(name=SECTION, status=status, msg=msg)
+    msg = "\n".join(
+        [
+            f"{res.path} -- {'ERROR' if res.errored else 'OK'}"
+            for res in lint_results
+        ],
+    )
+    has_errors = any(map(lambda res: res.errored != 0, lint_results))
+
+    return plug.Result(
+        name=SECTION,
+        msg=msg,
+        status=plug.Status.SUCCESS if not has_errors else plug.Status.ERROR,
+        data={
+            str(pylint_res.path): pylint_res.output
+            for pylint_res in lint_results
+        },
+    )
 
 
-def _pylint(python_files: Iterable[Union[pathlib.Path]]) -> Tuple[str, str]:
-    """Run ``pylint`` on all of the specified files.
+def _pylint(basedir: pathlib.Path) -> List[_PylintResult]:
+    """Run ``pylint`` on all python files in the specified directory.
 
     Args:
-        python_files: paths to ``.py`` files.
+        basedir: The base directory to search for files in.
     Returns:
-        (status, msg), where status is e.g. :py:const:`plugin.ERROR` and
-        the message describes the outcome in plain text.
+        A list of results from running pylint.
     """
+    python_files = list(basedir.rglob("*.py"))
+
     linted_files = []
     for py_file in python_files:
-        plug.echo("Running pylint on {!s}".format(py_file))
-        command = "pylint {!s}".format(py_file).split()
-        proc = subprocess.run(
+        command = shlex.split(f"pylint {py_file}")
+
+        full_lint = subprocess.run(
             command, stdout=subprocess.PIPE, stderr=subprocess.PIPE
         )
-
-        outfile = pathlib.Path(
-            "{}/{}.lint".format(py_file.parent, py_file.name)
+        error_lint = subprocess.run(
+            command + ["--errors-only"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
         )
-        outfile.touch()
-        outfile.write_bytes(proc.stdout)
-        linted_files.append(str(py_file))
 
-    msg = "linted files: {}".format(", ".join(linted_files))
-    return plug.Result(name=SECTION, status=plug.Status.SUCCESS, msg=msg)
+        output = full_lint.stdout.decode(sys.getdefaultencoding())
+        errored = error_lint.returncode != 0
+
+        linted_files.append(
+            _PylintResult(
+                path=py_file.relative_to(basedir),
+                output=output,
+                errored=errored,
+            )
+        )
+
+    return linted_files

--- a/tests/new_integration_tests/test_repos.py
+++ b/tests/new_integration_tests/test_repos.py
@@ -353,7 +353,7 @@ class TestClone:
             pylint_result, *_ = plug_results[repo_name]
             assert pylint_result.name == "pylint"
             assert pylint_result.status == plug.Status.SUCCESS
-            assert "src/main.py: OK" in pylint_result.msg
+            assert "src/main.py -- OK" in pylint_result.msg
 
     def test_pylint_plugin_with_python_syntax_error(
         self, platform_url, tmp_path_factory
@@ -379,7 +379,7 @@ class TestClone:
             pylint_result, *_ = plug_results[repo_name]
             assert pylint_result.name == "pylint"
             assert pylint_result.status == plug.Status.ERROR
-            assert "src/main.py: ERROR" in pylint_result.msg
+            assert "src/main.py -- ERROR" in pylint_result.msg
 
     def _setup_task_with_faulty_python_code(self, platform_url, workdir):
         python_task = workdir / "python-task"

--- a/tests/new_integration_tests/test_repos.py
+++ b/tests/new_integration_tests/test_repos.py
@@ -324,6 +324,7 @@ class TestClone:
         plug_results = funcs.run_repobee(
             f"repos clone -a {java_task.name} --base-url {platform_url} ",
             plugins=[_repobee.ext.javac],
+            workdir=workdir,
         )
 
         assert plug_results
@@ -343,6 +344,7 @@ class TestClone:
         plug_results = funcs.run_repobee(
             f"repos clone -a {python_task.name} --base-url {platform_url}",
             plugins=[_repobee.ext.pylint],
+            workdir=workdir,
         )
 
         assert plug_results
@@ -351,6 +353,47 @@ class TestClone:
             pylint_result, *_ = plug_results[repo_name]
             assert pylint_result.name == "pylint"
             assert pylint_result.status == plug.Status.SUCCESS
+            assert "src/main.py: OK" in pylint_result.msg
+
+    def test_pylint_plugin_with_python_syntax_error(
+        self, platform_url, tmp_path_factory
+    ):
+        """Test that the pylint plugin correctly reports errors."""
+        workdir = tmp_path_factory.mktemp("workdir")
+        python_task = self._setup_task_with_faulty_python_code(
+            platform_url, workdir
+        )
+        repo_names = plug.generate_repo_names(
+            STUDENT_TEAMS, [python_task.name]
+        )
+
+        plug_results = funcs.run_repobee(
+            f"repos clone -a {python_task.name} --base-url {platform_url}",
+            plugins=[_repobee.ext.pylint],
+            workdir=workdir,
+        )
+
+        assert plug_results
+        assert len(plug_results) == len(repo_names)
+        for repo_name in repo_names:
+            pylint_result, *_ = plug_results[repo_name]
+            assert pylint_result.name == "pylint"
+            assert pylint_result.status == plug.Status.ERROR
+            assert "src/main.py: ERROR" in pylint_result.msg
+
+    def _setup_task_with_faulty_python_code(self, platform_url, workdir):
+        python_task = workdir / "python-task"
+        python_code = (
+            "print('Hello, world!'"  # note missing closing parenthesis
+        )
+        create_local_repo(python_task, [("src/main.py", python_code)])
+        funcs.run_repobee(
+            f"repos setup -a {python_task.name} "
+            f"--base-url {platform_url} "
+            "--allow-local-templates",
+            workdir=workdir,
+        )
+        return python_task
 
     def _setup_task_with_python_code(self, platform_url, workdir):
         python_task = workdir / "python-task"

--- a/tests/new_integration_tests/test_repos.py
+++ b/tests/new_integration_tests/test_repos.py
@@ -8,6 +8,7 @@ import git
 import pytest
 
 import _repobee.ext.javac
+import _repobee.ext.pylint
 
 import repobee_plug as plug
 from repobee_testhelpers import localapi
@@ -331,6 +332,37 @@ class TestClone:
             javac_result, *_ = plug_results[repo_name]
             assert javac_result.name == "javac"
             assert javac_result.status == plug.Status.SUCCESS
+
+    def test_pylint_plugin_happy_path(self, platform_url, tmp_path_factory):
+        workdir = tmp_path_factory.mktemp("workdir")
+        python_task = self._setup_task_with_python_code(platform_url, workdir)
+        repo_names = plug.generate_repo_names(
+            STUDENT_TEAMS, [python_task.name]
+        )
+
+        plug_results = funcs.run_repobee(
+            f"repos clone -a {python_task.name} --base-url {platform_url}",
+            plugins=[_repobee.ext.pylint],
+        )
+
+        assert plug_results
+        assert len(plug_results) == len(repo_names)
+        for repo_name in repo_names:
+            pylint_result, *_ = plug_results[repo_name]
+            assert pylint_result.name == "pylint"
+            assert pylint_result.status == plug.Status.SUCCESS
+
+    def _setup_task_with_python_code(self, platform_url, workdir):
+        python_task = workdir / "python-task"
+        python_code = "print('Hello, world!')"
+        create_local_repo(python_task, [("src/main.py", python_code)])
+        funcs.run_repobee(
+            f"repos setup -a {python_task.name} "
+            f"--base-url {platform_url} "
+            "--allow-local-templates",
+            workdir=workdir,
+        )
+        return python_task
 
     def _setup_task_with_java_code(
         self, platform_url, workdir

